driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -101,135 +116,102 @@ public InternalConnection get() {
-            openAsync(connection, errHandlingCallback);
+        if (immediateConnection != null) {
+            openAsync(immediateConnection, eventSendingCallback, timeout);
         } else {
-            final long startTimeMillis = System.currentTimeMillis();
-            getAsyncGetter().submit(new Runnable() {
-                @Override
-                public void run() {
-                    try {
-                        if (getRemainingWaitTime() <= 0) {
-                            errHandlingCallback.onResult(null, createTimeoutException());
-                        } else {
-                            PooledConnection connection = getPooledConnection(getRemainingWaitTime(), MILLISECONDS);
-                            openAsync(connection, errHandlingCallback);
-                        }
-                    } catch (MongoTimeoutException e) {
-                        Exception exception = new MongoTimeoutException(format("Timeout waiting for a pooled connection after %d %s",
-                                settings.getMaxWaitTime(MILLISECONDS), MILLISECONDS));
-                        emitCheckOutFailedEvent(exception);
-                        errHandlingCallback.onResult(null, exception);
-                    } catch (Throwable t) {
-                        emitCheckOutFailedEvent(t);
-                        errHandlingCallback.onResult(null, t);
-                    }
+            getAsyncGetter().execute(() -> {
+                if (timeout.expired()) {
+                    eventSendingCallback.onResult(null, createTimeoutException(timeout));
+                    return;
                 }
-
-                private long getRemainingWaitTime() {
-                    return startTimeMillis + settings.getMaxWaitTime(MILLISECONDS) - System.currentTimeMillis();
+                PooledConnection connection;
+                try {
+                    connection = getPooledConnection(timeout);
+                } catch (RuntimeException e) {
+                    eventSendingCallback.onResult(null, e);
+                    return;
                 }
+                openAsync(connection, eventSendingCallback, timeout);
             });
         }
     }
 
-    private void emitCheckOutFailedEvent(final Throwable t) {
+    /**
+     * Sends {@link ConnectionCheckOutFailedEvent}
+     * and returns a {@link RuntimeException} that is reported as the cause of a checkout failure.
+     */
+    private RuntimeException checkOutFailed(final Throwable t) {
+        Throwable result = t;
         if (t instanceof MongoTimeoutException) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.TIMEOUT));
+        } else if (t instanceof MongoOpenConnectionInternalException) {
+            connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.CONNECTION_ERROR));
+            result = t.getCause();
+            assert result != null;
```

Comment:
`MongoOpenConnectionInternalException` is used for signalling that specifically `PooledConnection.open`/`openAsync` failed (this requires reporting `Reason.CONNECTION_ERROR`). As a result, I have to unwrap it here, because this exception is not part of our API.

While wrapping the cause in `MongoOpenConnectionInternalException` and then unwrapping it is not pretty, it allowed me to have only 2 places (1 call per `DefaultConnectionPool.get`/`getAsync` method)  in which `checkOutFailed` is called, instead of having it (`emitCheckOutFailedEvent`) called in 4 places plus having `new ConnectionCheckOutFailedEvent` created explicitly (outside of  `emitCheckOutFailedEvent`) in 3 more places. I think this is a good trade-off that simplified the code.